### PR TITLE
crimson/os/seastore/journal: allow pending i/o in a full record_submitter when rolling the segment

### PR DESCRIPTION
**Pull request by xxhdx1985126**
Fixes: https://tracker.ceph.com/issues/58824
Signed-off-by: Xuehan Xu <xxhdx1985126@gmail.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by cyx1231st** [comment]
Hmm, the fix direction is not reasonable to me.

Logically, it should be reasonable to allow to roll to a new segment for the incoming new i/o, even when there are still on-going i/os in the rolled segment, right? This was by-design -- rolling don't need to be blocked for this reason.

Probably the issue is elsewhere?

**Comment by xxhdx1985126** [comment]
> Probably the issue is elsewhere?

No, the issue was the assert at https://github.com/ceph/ceph/blob/acedd169e42d5b579ef303bc00e387898d9277eb/src/crimson/os/seastore/journal/segment_allocator.cc#L471-L473.

I know by-design journal allows on-going i/o when rolling segments, but this assert is not following that philosophy. When the `RecordSubmitter` is `FULL`, and there's pending i/o requests in it, this assert will be triggered.

And it seems, from the following code, that the original design was to wait for on-going i/o to finish before start rolling a segment:
https://github.com/ceph/ceph/blob/acedd169e42d5b579ef303bc00e387898d9277eb/src/crimson/os/seastore/journal/segment_allocator.cc#L475-L478

Am I missing anything?

**Comment by xxhdx1985126** [comment]
@cyx1231st Ah, sorry, the name for this pr is not appropriate. The segment rolling is not waiting for i/o, but is waiting for RecordSubmitter to be non-full. I'll change the name.

**Comment by cyx1231st** [comment]
I think the issue implies that the invariant is broken that in each do-write iteration, the submitter's io-depth should be increased at most 1.

https://github.com/ceph/ceph/blob/fce4edc8340b26acc937f5eb5d7dbb4be8e41d16/src/crimson/os/seastore/extent_placement_manager.cc#L100-L112

**Comment by xxhdx1985126** [comment]
> I think the issue implies that the invariant is broken that in each do-write iteration, the submitter's io-depth should be available and be increased at most 1, atomically.
> 
> https://github.com/ceph/ceph/blob/fce4edc8340b26acc937f5eb5d7dbb4be8e41d16/src/crimson/os/seastore/extent_placement_manager.cc#L100-L112

Yes, you are right. I think I've found the root cause. Right now, `RecordSubmitter::is_available()` checks whether `wait_available_promise` has value, and `wait_available_promise` has value only in the following two cases:

1. segment is being rolled;
2. `RecordSubmitter` is `FULL` and there are pending i/o records that `needs_flush`;

So when `RecordSubmitter` is `FULL` and there are pending i/o records that does NOT `needs_flush`, `RecordSubmitter` is still available:-)

**Comment by xxhdx1985126** [comment]
@cyx1231st After a second thought, I think maybe `RecordSubmitter`'s `unavailable` should be considered as `RecordSubmitter` is FULL and the `p_current_batch` needs flush, which is exactly what the current main branch code is doing.

Because if we consider `RecordSubmitter`'s `unavailable` as just `RecordSubmitter` being FULL, we have to first fill the `p_current_batch` before submitting the next io when `RecordSubmitter` recovers from being FULL, which seems unnecessary, as the filling of `p_current_batch` can perfectly be performed while the `RecordSubmitter` is FULL. What do you think?

**Comment by xxhdx1985126** [comment]
And from the following code, the original design seems to allow `RecordSubmitter` being FULL and `p_current_batch` is pending at the time of rolling segment, it just needed to wait for `RecordSubmitter` to recover. So it seems that the original code is intentionally considering `RecordSubmitter`'s `unavailable` as `RecordSubmitter` being FULL and `p_current_batch` needing flush.

https://github.com/ceph/ceph/blob/acedd169e42d5b579ef303bc00e387898d9277eb/src/crimson/os/seastore/journal/segment_allocator.cc#L473-L479

**Comment by cyx1231st** [comment]
I think the seg-fault is a representation of a logic loophole that when `submit(record_t)` and `roll_segment()` are both called atomically, the following corner case is not handled properly, with contradictory asserts and logic:
https://github.com/ceph/ceph/blob/fce4edc8340b26acc937f5eb5d7dbb4be8e41d16/src/crimson/os/seastore/journal/segment_allocator.cc#L572-L581
and
https://github.com/ceph/ceph/blob/fce4edc8340b26acc937f5eb5d7dbb4be8e41d16/src/crimson/os/seastore/journal/segment_allocator.cc#L469-L472

**Comment by xxhdx1985126** [comment]
> I think the seg-fault is a representation of a logic loophole that when `submit(record_t)` and `roll_segment()` are both called atomically, the following corner case is not handled properly, with contradictory asserts and logic

Yes, this PR is about this `contradictory asserts and logic`.

BTW, which seg-fault are you refering to?

File: src/crimson/os/seastore/journal/segment_allocator.h [hunk 1]
```diff
@@ -408,10 +410,15 @@ class RecordSubmitter {
...
+  // For the time being, there are only two possible sources for the
+  // record submitter to be unavailable:
+  //    1. RecordSubmitter is "FULL" and the current batch needs flush
+  //    2. Segment of the underlying segment allocator is being rolled
+  size_t unavailable_sources = 0;
   bool has_io_error = false;
   // when needs flush but io depth is full,
   // wait for decrement_io_with_flush()
-  std::optional<seastar::promise<> > wait_unfull_flush_promise;
+  std::optional<seastar::shared_promise<> > wait_unfull_flush_promise;
```
Comment (cyx1231st) [inline src/crimson/os/seastore/journal/segment_allocator.h:421]: Hmm, I think we may not need to generalize wait_available with sources, or relax the wait_unfull_flush_promise to be a shared promise, just to correct a corner case.

How do you think of https://github.com/ceph/ceph/pull/50273/commits/9f8bd71815f5a31c54c922a56225bde072dfdf3b ?

I also include you as the author.

**Comment by xxhdx1985126** [comment]
closing this PR in favour of #50273

**Pull request closed without merging**